### ISSUE TITLE
Use `client-go` to make requests to kube-apiserver instead of direct API calls

### ISSUE THREAD

**Issue by daemon1024** [kubearmor/KubeArmor#531]
## Feature Request

Use `client-go` to make requests to kube-apiserver instead of direct API calls

**Is your feature request related to a problem? Please describe the use case.**

Currently we need to have a running proxy for kube-apiserver for KubeArmor in process mode to interact with k8s. 

**Describe the solution you'd like**

Current : https://github.com/kubearmor/KubeArmor/blob/60c105f04f1f3bf6fdc5d90264236e104c153d34/KubeArmor/core/k8sHandler.go#L238

Proposing: 
```
nodeWatcher, err := clientset.CoreV1().Nodes().Watch(metav1.ListOptions{})
```
Ref https://github.com/kubernetes/client-go#how-to-use-it

**Comment by afzal442** [comment]
hey @daemon1024, I got that proposed way to refactor that as you mentioned abv but unsure about incluster perspective.
Let me know how you would send response after getting `nodeWatcher` as example. Thanks.

// @nyrahul

**Comment by daemon1024** [comment]
Hey @afzal442, Thanks for taking interest in this issue.

> but unsure about incluster perspective.

As per my understanding, only difference is that how you authenticate with the server, rest the API is the same. The authentication part is already handled in https://github.com/kubearmor/KubeArmor/blob/main/KubeArmor/core/k8sHandler.go#L97-L102. So it's not an issue.

> Let me know how you would send response after getting nodeWatcher as example.

You can take a look at https://github.com/iximiuz/client-go-examples/tree/main/watch-typed-simple for inspiration. 
Basically we make repeated requests inside a for loop right now. Instead we can loop the channel provided by nodeWatcher.

If you have more queries feel free to ping me in this thread or on our [slack channel](https://kubearmor.herokuapp.com/) ✌🏽

**Comment by sbin64** [comment]
Thanks so much, @daemon1024! That makes sense now. 

> You can take a look at https://github.com/iximiuz/client-go-examples/tree/main/watch-typed-simple for inspiration.

That's really inspired me, that's why I landed on here. 😄 
That is written wrt outside of cluster perspective.

 > Instead we can loop the channel provided by `nodeWatcher`.

```
go func() {
		for event := range nodeWatcher.ResultChan() {
			fmt.Printf(
				"Watch Event: %s %s\n",
				event.Type, event.Object.GetObjectKind().GroupVersionKind().Kind,
			)
		}
	}()
```
I'm working around here! I'll ping you on slack soon. Thanks

**Comment by sbin64** [comment]
After going through your proposal and suggestion made, have a look  for nodeWatcher as below:
changes at k8sHandler.go code
```
// WatchK8sNodes Function
func (kh *K8sHandler) WatchK8sNodes() watch.Interface {
	if !kl.IsK8sEnv() { // not Kubernetes
		return nil
	}

	if kl.IsInK8sCluster() { // kube-apiserver

		if kh.InitInclusterAPIClient() {
			watchNodes, err := kh.K8sClient.
				CoreV1().
				Nodes().
				Watch(
					context.Background(),
					metav1.ListOptions{},
				)
			if err != nil {
				return nil
			}

			return watchNodes
		}

	}

	// kube-proxy (local)
	if kh.InitLocalAPIClient() {
		watchNodes, err := kh.K8sClient.
			CoreV1().
			Nodes().
			Watch(
				context.Background(),
				metav1.ListOptions{},
			)

		if err != nil {
			return nil
		}

		return watchNodes
	}
	return nil
}
```
changes at kubeUpdate.go as below
```
// WatchK8sNodes Function
func (dm *KubeArmorDaemon) WatchK8sNodes() {
	for {
		if nodeWatcher := K8s.WatchK8sNodes(); nodeWatcher != nil {
			// defer nodeWatcher.Close()

			go func() {
				for event := range nodeWatcher.ResultChan() {

					// Kubearmor uses hostname to get the corresponding node information, but there are exceptions.
					// For example, the node name on EKS can be of the format <hostname>.<region>.compute.internal
					nodeName := strings.Split(event.Object.ObjectMeta.Name, ".")
					if nodeName[0] != cfg.GlobalCfg.Host {
						continue
					}

					node := tp.Node{}

					for _, address := range event.Object.Status.Addresses {
						if address.Type == "InternalIP" {
							node.NodeIP = address.Address
							break
						}
					}

					node.Annotations = map[string]string{}
					node.Labels = map[string]string{}
					node.Identities = []string{}

					// update annotations
					for k, v := range event.Object.ObjectMeta.Annotations {
						node.Annotations[k] = v
					}

					// update labels and identities
					for k, v := range event.Object.ObjectMeta.Labels {
						node.Labels[k] = v
						node.Identities = append(node.Identities, k+"="+v)
					}

					sort.Slice(node.Identities, func(i, j int) bool {
						return node.Identities[i] < node.Identities[j]
					})

					// node info
					node.Architecture = event.Object.Status.NodeInfo.Architecture
					node.OperatingSystem = event.Object.Status.NodeInfo.OperatingSystem
					node.OSImage = event.Object.Status.NodeInfo.OSImage
					node.KernelVersion = event.Object.Status.NodeInfo.KernelVersion
					node.KubeletVersion = event.Object.Status.NodeInfo.KubeletVersion

					// container runtime
					node.ContainerRuntimeVersion = event.Object.Status.NodeInfo.ContainerRuntimeVersion

					dm.HandleNodeAnnotations(&node)

					dm.Node = node
				}
			}()
		} else {
			time.Sleep(time.Second * 1)
		}
	}
}
```

**Comment by daemon1024** [comment]
Looks in line with what I was thinking. Can you put it in a draft PR it's easy to review things here.

Are you able to compile and run things?

**Comment by nyrahul** [comment]
folks, any update on this issue? Thanks

**Comment by Chinwendu20** [comment]
Any update? Can I work on this?

**Comment by daemon1024** [comment]
Partially handled in #1032
